python/ml4ir/applications/ranking/tests/test_feature_layer.py

```diff
@@ -686,3 +688,30 @@ def test_global_1d_pooling(self):
         except KeyError:
             found_key_error = True
         assert found_key_error
+
+    def test_tf_native_op(self):
+        """
+        Unit test the tf_native_op feature function
+
+        Checks the right output shapes produced and the values generated
+        """
+        input_tensor = np.abs(np.random.randn(32, 4, 2))
+
+        actual_tensor = tf_native_fns.tf_native_op(
+                feature_tensor=input_tensor,
+                feature_info={
+                    "name": "f",
+                    "feature_layer_info": {
+                        "args": {
+                            "ops": [
+                                {"fn": "tf.math.add", "kwargs": {"y": 1.}},
+                                {"fn": "tf.math.log"}
+                            ]
+                        }
+                    }
+                },
+                file_io=None
+            )
+        expected_tensor = tf.expand_dims(tf.math.log(tf.math.add(input_tensor, 1.)), axis=-1)
```

Comment:
why are we `expand_dims` here? batch?

Reply:
Even I was confused why I had to do it, then I realized that we have an `expand_dims` in feature_layer, but it is gated with those features that do not have a feature fn specified -> https://github.com/salesforce/ml4ir/blob/master/python/ml4ir/base/features/feature_layer.py#L171

I did try removing the `if` condition, but it breaks every other feature fn like embedding, LSTM etc. I will need to modify everyone of them. I didn't want to devolve this PR into that at the time being. So I am adhering to the `expand_dims` we have in other functions as well.

Reply:
We do it there to support for ranking type of datasets, where a query can be coupled with several retrieved docs (if I remember well).. Yes, it is a tough thing to remember that one, it is everywhere, even in the metrics..

Reply:
Yes, that is the reason.